django/http/response.py

```diff
@@ -15,13 +16,75 @@
 from django.core.serializers.json import DjangoJSONEncoder
 from django.http.cookie import SimpleCookie
 from django.utils import timezone
+from django.utils.datastructures import CaseInsensitiveMapping
 from django.utils.encoding import iri_to_uri
 from django.utils.http import http_date
 from django.utils.regex_helper import _lazy_re_compile
 
 _charset_from_content_type_re = _lazy_re_compile(r';\s*charset=(?P<charset>[^\s;]+)', re.I)
 
 
+class ResponseHeaders(CaseInsensitiveMapping):
+    def __init__(self, data):
+        """
+        Populate the initial data.
+
+        Works similarly to ``CaseInsensitiveMapping`` but sets the items
+        directly to ensure the values are correctly encoded.
+        """
+        if not isinstance(data, Mapping):
+            data = {
+                k: v
+                for k, v in CaseInsensitiveMapping._destruct_iterable_mapping_values(data)
+            }
+        self._store = {}
+        for header, value in data.items():
+            self[header] = value
+
+    def _convert_to_charset(self, value, charset, mime_encode=False):
+        """
+        Convert headers key/value to ascii/latin-1 native strings.
+
+        ``charset`` must be 'ascii' or 'latin-1'. If ``mime_encode`` is True
+        and ``value`` can't be represented in the given charset, apply
+        MIME-encoding.
+        """
+        if not isinstance(value, (bytes, str)):
+            value = str(value)
+        if ((isinstance(value, bytes) and (b'\n' in value or b'\r' in value)) or
+                isinstance(value, str) and ('\n' in value or '\r' in value)):
+            raise BadHeaderError("Header values can't contain newlines (got %r)" % value)
+        try:
+            if isinstance(value, str):
+                # Ensure string is valid in given charset
+                value.encode(charset)
+            else:
+                # Convert bytestring using given charset
+                value = value.decode(charset)
+        except UnicodeError as e:
+            if mime_encode:
+                value = Header(value, 'utf-8', maxlinelen=sys.maxsize).encode()
+            else:
+                e.reason += ', HTTP response headers must be in %s format' % charset
+                raise
+        return value
+
+    def __delitem__(self, key):
+        del self._store[key.lower()]
```

Comment:
This isn't equivalent to the old implementation.

https://github.com/django/django/blob/9369f0cebba1f65909a14dec6aa3515ec1eb2557/django/http/response.py#L140-L141

From the docs: 

> Note that unlike a dictionary, del doesn’t raise KeyError if the header field doesn’t exist.

Can you add tests asserting that `del response.headers["not-there"]` does not raise. 

(GZip Middleware tests will fail pending this change.)

Reply:
That's the implementation of `__delitem__` on the store itself. On the response it look like this:

```
def __delitem__(self, header):
        self.headers.pop(header, False)
```

This does lead to a bit of a discrepancy between the two ways of handling them, I'm not sure if it's better to have it like this, where it's more like a regular dict, or to have the two interfaces as similar as possible. What do you think?

Reply:
I think the interfaces should be the same. It's probably the right (more forgiving) behaviour for a response, which isn't a dict. Folks will `s/response[/response.headers[/` not the least. But the docs also say this is the behaviour.

Reply:
I've sorted that now and added a test for it.